libs/MeshKernel/src/SplitRowColumnOfMesh.cpp

```diff
@@ -0,0 +1,350 @@
+        }
+    }
+    else if (IsQuadrilateral(mesh, previousElementId))
+    {
+        if (!IsValid(nextElementId) || IsQuadrilateral(mesh, nextElementId))
+        {
+            // This is probably the most common case in a sequence of quadrilateral elements
+            UInt oppositeEdgeId = OppositeEdgeId(mesh, elementId, edgeId);
+            UInt secondNewNodeId = SplitEdge(mesh, oppositeEdgeId, edgesToDelete, undoActions);
+
+            std::tie(std::ignore, undoActions.Insert()) = mesh.ConnectNodes(newNode, secondNewNodeId);
+
+            // Overwrite newNode with the latest new node id.
+            newNode = secondNewNodeId;
+        }
+        else if (!IsQuadrilateral(mesh, nextElementId))
+        {
+            UInt oppositeEdgeId = OppositeEdgeId(mesh, elementId, edgeId);
+            const Edge oppositeEdgeNode = mesh.GetEdge(oppositeEdgeId);
+
+            std::tie(std::ignore, undoActions.Insert()) = mesh.ConnectNodes(oppositeEdgeNode.first, newNode);
+            std::tie(std::ignore, undoActions.Insert()) = mesh.ConnectNodes(newNode, oppositeEdgeNode.second);
+
+            // Set the newNode with the null id, indicating no new node was created
+            newNode = constants::missing::uintValue;
+        }
+    }
+    else if (!IsQuadrilateral(mesh, previousElementId))
+    {
+        if (!IsValid(nextElementId) || IsQuadrilateral(mesh, nextElementId))
+        {
+            UInt oppositeEdgeId = OppositeEdgeId(mesh, elementId, edgeId);
+            UInt secondNewNodeId = SplitEdge(mesh, oppositeEdgeId, edgesToDelete, undoActions);
+
+            std::tie(std::ignore, undoActions.Insert()) = mesh.ConnectNodes(edgeNode.first, secondNewNodeId);
+            std::tie(std::ignore, undoActions.Insert()) = mesh.ConnectNodes(secondNewNodeId, edgeNode.second);
+
+            // Overwrite newNode with the latest new node id.
+            newNode = secondNewNodeId;
+        }
+
+        // If neither the previous nor the next elements are quadrilaterals then there is nothing to do.
+    }
+}
+
+bool meshkernel::SplitRowColumnOfMesh::IsQuadrilateral(const Mesh2D& mesh, const UInt elementId) const
+{
+    return IsValid(elementId) && mesh.m_numFacesNodes[elementId] == constants::geometric::numNodesInQuadrilateral;
+}
+
+void meshkernel::SplitRowColumnOfMesh::CollectElementsToSplit(const Mesh2D& mesh, const UInt edgeId, std::vector<UInt>& elementIds, std::vector<UInt>& edgeIds) const
+{
+    const UInt firstElementId = mesh.m_edgesFaces[edgeId][0] != constants::missing::uintValue ? mesh.m_edgesFaces[edgeId][0] : mesh.m_edgesFaces[edgeId][1];
+    const UInt initialNumberOfElements = static_cast<UInt>(std::sqrt(static_cast<double>(mesh.GetNumNodes())));
+
+    std::vector<UInt> partialElementIds;
+    std::vector<UInt> partialEdgeIds;
+
+    partialElementIds.reserve(initialNumberOfElements);
+    partialEdgeIds.reserve(initialNumberOfElements);
+
+    // Loop over each side of the edge
+    for (UInt i = 0; i < 2; ++i)
```

Comment:
create a function and call it twice? this could reduce the cyclomatic complexity a bit